### PR TITLE
Improve signal safety

### PR DESCRIPTION
**Pull request by N-R-K**
The purpose of this patch-set is to improve entr's [signal safety](https://man7.org/linux/man-pages/man7/signal-safety.7.html).

More detailed information are available on each commit message.

(Also includes a small indentation fix).

### CODE REVIEW

**Comment by N-R-K** [comment]
Just noticed there's also `handle_exit()` which calls `killpg` and `exit` (neither of which are declared as async-signal-safe).

**Comment by N-R-K** [comment]
Switched to `_exit` in `handle_exit` as well. Left `killpg` as is, it's functionally similar to a subset of `kill` - which is async-signal-safe. And so `killpg` should be fine in practice as well.

The fact that killpg is not listed as async-signal-safe is probably an oversight by the POSIX committee, I will most likely file a bug against it.

**Comment by eradman** [comment]
My understanding is that `exit(3)` is the function to call for "normal program termination" because it does cleanup. I think `_exit(2)` is for cases where you do not want buffers to be flushed,  tempfiles unlinked, and so on.

So a signal handler in a child process would be a good candidate since you probably don't want a child process to do any cleanup.

**Comment by N-R-K** [comment]
> I think `_exit(2)` is for cases where you do not want buffers to be flushed

That's correct, it also skips `atexit` (though entr doesn't use `atexit` so not very relavant).

And it's not sound to do anything with a FILE stream from a signal handler. The [linux manpage](https://man7.org/linux/man-pages/man7/signal-safety.7.html) already gives one example, another example would be the following (keep in mind `stdio` functions typically will use locks so that they can be thread-safe):

1. Stdio function locks mutex M associated with `stdout`.
2. Signal arrives.
3. Signal-handler tries to call a stdio function on `stdout` (e.g `printf`).
4. stdio function tries to lock mutex M, sees it's already locked and starts waiting for it to unlock.

In other words, you get a deadlock. (And this can in fact happen in practice, I've seen a dozen programs where user's report randomly hanging/crashing and the culprit was doing non-signal-safe things inside a signal handler). And this is just one possible way how things can go wrong if stdio functions are called from a signal-handler.

In short: signal handlers need to be careful not to call anything that isn't listed as `async-signal-safe`.

**Comment by eradman** [comment]
I can see how a signal handler that is not async-safe could end terribly (`SIGABRT`, mangled output), but I would be surprised if functions in the standard C library could deadlock. I'll see if I can find bug reports in other projects that indicate this.

With this change is something odd when running the system tests. @N-R-K can you look into this?

```
$ make check
./system_test.sh
...............
FAIL: exec a command using one-shot and shell options and return signal
'' != 'terminated by signal 9'
make: *** [Makefile.bsd:12: check] Error 1
$ uname -a
Linux nuc-5250u 5.19.0-40-generic #41~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Fri Mar 31 16:00:14 UTC 2 x86_64 x86_64 x86_64 GNU/Linux

```

**Comment by N-R-K** [comment]
> I'll see if I can find bug reports in other projects that indicate this.

Here's one from `openrc` due to calling malloc/free inside signal-handler: https://github.com/OpenRC/openrc/issues/589
And a crash report in `nsxiv` caused by the signal handler spoiling the `errno`: https://codeberg.org/nsxiv/nsxiv/issues/391

I can give more examples if you'd like.

> With this change is something odd when running the system tests. @N-R-K can you look into this?

Sure, I'll take a look into in a bit.

**Comment by N-R-K** [comment]
> With this change is something odd when running the system tests. @N-R-K can you look into this?

Found the issue, in commit https://github.com/eradman/entr/pull/118/commits/767837fe1041ff5ac2851613db0de3ab05da93af I mistakenly wrote the message into `stderr` rather than `stdout` (since `stderr` is typically more common for messages like these).

Fixed now. All tests are passing.

**Comment by eradman** [comment]
Good, tests pass on Linux, OpenBSD and MacOS

Please drop commit 17b285a "small indentation fix". There are some changes to indentation that I'll look at separately. I have had the habit of indenting a section with a tab, but doing line continuations with four spaces. Mixing them in this way is confusing to people.

Is there any advantage to use `snprintf(3)/write(2)` instead of `fprintf(3)`? Either way it's not technically correct, and neither approach should cause a problem for users.

**Comment by N-R-K** [comment]
I've dropped the indentation changes.

> Is there any advantage to use `snprintf(3)/write(2)` instead of `fprintf(3)`?

Yes, it's not safe to call `fprintf` from a signal handler. From [signal-safety manpage](https://man7.org/linux/man-pages/man7/signal-safety.7.html):

> The kinds of issues that render a function unsafe can be quickly understood when one considers the implementation of the stdio library, all of whose functions are not async-signal-safe.
>
> When performing buffered I/O on a file, the stdio functions must maintain a statically allocated data buffer along with associated counters and indexes (or pointers) that record the amount of data and the current position in the buffer.  Suppose that the main program is in the middle of a call to a stdio function such as [printf(3)](https://man7.org/linux/man-pages/man3/printf.3.html) where the buffer and associated variables have been partially updated.  If, at that moment, the program is interrupted by a signal handler that also calls [printf(3)](https://man7.org/linux/man-pages/man3/printf.3.html), then the second call to [printf(3)](https://man7.org/linux/man-pages/man3/printf.3.html) will operate on inconsistent data, with unpredictable results.

File: entr.c [hunk 1]
```diff
@@ -159,6 +161,14 @@ main(int argc, char *argv[]) {
 	/* ensure a shell is available to use */
 	setenv("SHELL", "/bin/sh", 0);
 
+	shell = getenv("SHELL");
+	/* NOTE: used throughout the entire program, freed automatically at
+	 * process termination. */
```
Comment (eradman) [inline entr.c:166]: we can elide this comment

File: entr.c [hunk 2]
```diff
@@ -257,20 +269,28 @@ proc_exit(int sig) {
...
+			_exit(WEXITSTATUS(child_status));
 	}
+	/* restore errno so that the resuming code is unimpacted. */
+	errno = saved_errno;
 }
 
 void
 print_child_status(int status) {
+	int n;
+	char buf[1 << 11];
```
Comment (eradman) [inline entr.c:283]: Replace with `2048`

File: entr.c [hunk 3]
```diff
@@ -237,14 +249,14 @@ handle_exit(int sig) {
...
-	    exit(0);
+	    _exit(0);
 	else
 	    raise(sig);
 }
 
 void
 proc_exit(int sig) {
-	int status;
+	int status, saved_errno = errno;
```
Comment (eradman) [inline entr.c:259]: Not incorrect, but put `saved_errno` declaration on it's own line

**Review by eradman — CHANGES_REQUESTED** [review]

**Comment by eradman** [comment]
Thanks @N-R-K. I can't see any downside to these changes.

I'll probably tweak the comments before the next release
